pandas/io/common.py

```diff
@@ -191,14 +192,41 @@ def get_filepath_or_buffer(
 
         .. versionadded:: 1.2.0
 
+    ..versionchange:: 1.2.0
+
+      Returns `mode`. If a path-like object is converted to a file-like object, the
+      returned mode is binary.
+
     Returns
     -------
-    Tuple[FilePathOrBuffer, str, CompressionOptions, bool]
-        Tuple containing the filepath or buffer, the encoding, the compression
-        and should_close.
+    Tuple[FilePathOrBuffer, str, CompressionOptions, bool, Optional[str]]
+        Tuple containing the filepath or buffer, the encoding, the compression,
+        should_close, and the mode.
     """
     filepath_or_buffer = stringify_path(filepath_or_buffer)
 
+    # bz2 and xz do not write the byte order mark for utf-16 and utf-32
+    # print a warning when writing such files
+    compression_method = infer_compression(
+        filepath_or_buffer, get_compression_method(compression)[0]
+    )
+    if (
+        mode
+        and "w" in mode
+        and compression_method in ["bz2", "xz"]
+        and encoding in ["utf-16", "utf-32"]
+    ):
+        warnings.warn(
+            f"{compression} will not write the byte order mark for {encoding}",
+            UnicodeWarning,
+        )
+
+    # use binary mode when converting path-like objects to file-like objects
+    # except when text mode is explicitly requested
+    binary_mode = mode or "rb"
+    if "t" not in binary_mode and "b" not in binary_mode:
```

Comment:
does this change a simple mode='w' ?

Reply:
or are you not actually using this (except to return)?

Reply:
`binary_mode` is used to open `ffspec` files and is only returned in this case. In the case that `ffspec` is used and the caller had `mode="w"`, `mode="wb"` will be returned as `ffspec` files are now always opened in binary mode (unless mode contains a `"t"` to explicitly request text mode (`to_json` does that)). If `ffspec` isn't used to open a file, the initial mode is returned.

If we do not do that, users who want to write a compressed CSV file to the google cloud, need to specify `mode="wb"` manually. If we are fine with that, we wouldn't need to change the interface of `get_filepath_or_buffer`. But it would be a little bit un-intuitive to require to specify `mode` when a path/string is provided.

Probably should rename `binary_mode` to something like `ffspec_mode`.